File: dl.go
package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"time"
)

func main() {
	start := time.Now()
	defer func() {
		fmt.Printf("Total time: %.4fs\n", time.Since(start).Seconds())
	}()

	items := []string{
		"http://google.com",
		"http://python.org",
		"http://ruby-lang.org",
		"http://golang.org",
	}

	c := make(chan string)

	for _, item := range items {
		go count(item, c)
	}

	timeout := time.After(500 * time.Millisecond)
	for _ = range items {
		select {
		case result := <-c:
			fmt.Print(result)
		case <-timeout:
			return
		}
	}
}

func count(url string, c chan<- string) {
	start := time.Now()
	r, err := http.Get(url)
	if err != nil {
		panic(err)
	}
	n, err := io.Copy(ioutil.Discard, r.Body)
	if err != nil {
		panic(err)
	}
	r.Body.Close()
	dt := time.Since(start).Seconds()
	c <- fmt.Sprintf("%s %d [%.4fs]\n", url, n, dt)
}

File: dl.py
import multiprocessing
import urllib2
import Queue
import time


urls_to_load = [
    "http://google.com",
    "http://python.org",
    "http://ruby-lang.org",
    "http://golang.org",
]


limit = .5  # seconds


def read_url(url, queue):
    now = time.time()
    data = urllib2.urlopen(url).read()
    dt = time.time() - now
    queue.put((url, len(data), dt))


def fetch_parallel():
    start = time.time()
    queue = multiprocessing.Queue()
    processes = [multiprocessing.Process(target=read_url, args=(url, queue)) for url in urls_to_load]
    for p in processes:
        p.start()
    for p in processes:
        p.join(timeout=limit - (time.time() - start))
        if p.is_alive():
            p.terminate()
        if not queue.empty():
            print "%s %d [%.4fs]" % queue.get()
    print "Total time: %.4fs" % (time.time() - start)


if __name__ == "__main__":
    fetch_parallel()
